File: windows/app/src/main/kotlin/app/recly/windows/i18n/UiMessage.kt
package app.recly.windows.i18n

/**
 * A line for the user that is decided away from the screen — in [app.recly.windows.ui.ShellModel],
 * in the mapping of a `SyncResult` onto words, in a `step_run.last_error`.
 *
 * It stays a name until something draws it. The models outlive any window and the language can
 * change under them, so a sentence resolved when the state was built would still be in the old
 * language after the tray and the windows redrew (docs/07 rule 3); a key and its arguments cannot
 * go stale. This is the phone's `UiMessage` with a [Str] where its resource id is.
 *
 * The exception is text that came from somewhere else — the parser's own wording (docs/02), a core
 * diagnostic, a title the user typed — which is shown as it arrived.
 */
sealed interface UiMessage {
    /**
     * [args] are `String.format` arguments, so an `Int` stays an `Int` for a `%1$d`.
     *
     * One of them may be another [UiMessage] — a sentence with a core message inside it — and it is
     * resolved along with this one rather than before it, so nothing here goes stale either.
     */
    data class Res(val key: Str, val args: List<Any> = emptyList()) : UiMessage

    data class Text(val value: String) : UiMessage
}

/** A [UiMessage] as words, in whatever language [strings] is. */
fun UiMessage.text(strings: Strings): String = when (this) {
    is UiMessage.Res -> {
        // A nested message is resolved with this one, not before it, so it cannot go stale either.
        val resolved = args.map { if (it is UiMessage) it.text(strings) else it }
        strings.get(key, *resolved.toTypedArray())
    }

    is UiMessage.Text -> value
}

/** Shorthand for the common case of a key with no arguments. */
fun Str.message(vararg args: Any): UiMessage = UiMessage.Res(this, args.toList())

File: apple/RecKit/Sources/RecKit/Localization/UiMessage.swift
import Foundation

/// docs/07 rule 3: a message a model holds on to — a banner, the note an import left behind — as what
/// it *is* rather than as words.
///
/// Prose resolved when the model was written would outlive the language change the screen is meant
/// to answer: the banner is still standing, and it is still in the old language. So the model keeps
/// the key and its arguments, and [text] makes the sentence where the view draws it.
public enum UiMessage: Equatable, Sendable {
    /// A key in RecKit's own catalog, with the arguments the sentence takes. An argument is usually
    /// nothing to translate — a count, a secret name, a status, all [verbatim] — but it may itself
    /// be a message, which is how a core code becomes a sentence *inside* this one rather than a
    /// code left standing in the middle of it.
    case key(String, args: [UiMessage] = [])
    /// A core `last_error` / `CoreMessage` code (docs/07 §5), which [CoreMessages] turns into a
    /// sentence — and which an older build may have written as prose, passed through unchanged.
    case core(String)
    /// Text that has no translation at all: the parser's own list of complaints, a name the user
    /// typed.
    case verbatim(String)

    /// The sentence, in the app's language *now*.
    public var text: String {
        switch self {
        case .key(let key, let args):
            guard !args.isEmpty else { return RecKitStrings.localized(key) }
            return String(
                format: RecKitStrings.localized(key),
                locale: AppLanguage.locale,
                arguments: args.map(\.text) as [CVarArg]
            )
        case .core(let code):
            return CoreMessages.text(code).sentence
        case .verbatim(let text):
            return text
        }
    }

    /// docs/07 rule 4: logs are not localized. What identifies a message in a log is what it was
    /// stored as — the key, or the core's code — never [text], which is whichever language the
    /// device happened to be in when the line was written and changes under a reader.
    public var logCode: String {
        switch self {
        case .key(let key, _): return key
        case .core(let code): return code
        case .verbatim: return "verbatim"
        }
    }
}
